File: Data Structures & Algorithms/house-robber-ii/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        nums1 = nums[:n-1]
        dp1 = [-1] *len(nums1)
        nums2 = nums[1:]
        dp2 = [-1] * len(nums2)
        memo = [dp1,dp2]
        print(memo)
        
        def dfs(i, num,ch):
            if i >= len(num):
                return 0
            if memo[ch][i] == -1:
                memo[ch][i] = max(dfs(i+1,num,ch), dfs(i+2,num,ch) + num[i])
            return memo[ch][i]
        return max(dfs(0,nums1,0), dfs(0,nums2,1))

File: Data Structures & Algorithms/count-paths/submission-2.cpp
class Solution {
private: vector<vector<int>> memo;
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>> dp(m+1, vector(n+1,0));
        dp[m-1][n-1] = 1;
        for (int i=m-1;i>=0;i--) {
            for (int j=n-1;j>=0;j--) {
                dp[i][j] += dp[i+1][j] + dp[i][j+1];
            }
        }
        return dp[0][0];
    }
    int dfs(int i, int j , int m, int n) {
        if (i>=m || j >= n) {
            return 0;
        }
        if (i == m-1 && j == n-1) {
            return 1;
        }
        if (memo[i][j] != -1) {
            return memo[i][j];
        }
         return memo[i][j] = dfs(i+1, j, m ,n) + dfs(i,j+1,m,n);
    }
};
